docs/source/howto/data.rst

```diff
@@ -248,11 +247,247 @@ The node class has various methods to set these attributes, such as :py:`~aiida.
+#.  In the *queryhelp* approach, you construct a dictionary that defines your query and pass it to the ``QueryBuilder``.
+
+Both APIs provide the same functionality - the appender method may be more suitable for interactive use, e.g., in the ``verdi shell``, whereas the queryhelp method can be useful in scripting.
+In this how-to we will focus on the basics of the appender method.
+For more advanced queries or more details on the queryhelp, see the :ref:`topics section on advanced querying <topics:database:advancedquery>`.
+
+.. _how-to:data:find:select:
+
+Selecting entities
+------------------
+
+Using the ``append()`` method of the ``QueryBuilder``, you can query for the entities you are interested in.
+Suppose you want to query for calculation job nodes in your database:
+
+.. code-block:: python
+
+    from aiida.orm.querybuilder import QueryBuilder
+    qb = QueryBuilder()       # Instantiating instance. One instance -> one query
+    qb.append(CalcJobNode)    # Setting first vertex of path
+
+If you are interested in instances of different classes, you can also pass a tuple, list or set of classes.
+However, they have to be of the same ORM-type (e.g. all have to be subclasses of ``Node``):
+
+.. code-block:: python
+
+    qb = QueryBuilder()       # Instantiating instance. One instance -> one query
+    qb.append([CalcJobNode, WorkChainNode]) # Setting first vertice of path, either WorkChainNode or Job.
+
+.. note::
+
+    Remember that :ref:`processes<topics:processes:concepts:types>` have both a run-time ``Process`` that steers them and a ``Node`` that stores their metadata in the database.
+    The QueryBuilder allows you to pass either the ``Node`` class (e.g. ``CalcJobNode``) or the ``Process`` class (e.g. ``CalcJob``), which will automatically apply the correct filters for the type of calculation.
+
+.. _how-to:data:find:results:
+
+Retrieving results
+------------------
+
+Once you have *appended* the entity you want to query for to the ``QueryBuilder``, the next question is how to get the results.
+There are several ways to obtain data from a query:
+
+.. code-block:: python
+
+    qb = QueryBuilder()                 # Instantiating instance
+    qb.append(CalcJobNode)              # Setting first vertice of path
+
+    first_row = qb.first()              # Returns a list (!)
+                                        # of the results of the first row
+
+    all_results_d = qb.dict()           # Returns all results as
+                                        # a list of dictionaries
+
+    all_results_l = qb.all()            # Returns a list of lists
+
+In case you are working with a large dataset, you can also return your query as a generator:
+
+.. code-block:: python
+
+    all_res_d_gen = qb.iterdict()       # Return a generator of dictionaries
+                                        # of all results
+    all_res_l_gen = qb.iterall()        # Returns a generator of lists
```

Comment:
I think it would be good to add an example, yes. E.g.:

```
.. code-block:: python

    all_res_d_gen = qb.iterdict()       # Return a generator of dictionaries
                                        # of all results
    all_res_l_gen = qb.iterall()        # Returns a generator of lists

This will retrieve the data in batches, and you can start working with the data before the query has completely finished.
For example, you can iterate over the results of your query in a for loop:

.. code-block:: python

    for entry in builder.iterall():
        # do something with a single entry in the query result
```